File: platform/src/main/java/com/auraboot/framework/meta/service/FieldLibraryService.java
package com.auraboot.framework.meta.service;

import com.auraboot.framework.common.dto.PageResult;
import com.auraboot.framework.meta.dto.FieldRecommendation;
import com.auraboot.framework.meta.dto.FieldSearchRequest;
import com.auraboot.framework.meta.dto.MetaFieldDTO;

import java.util.List;
import java.util.Map;

/**
 * Field library service interface
 * Provides field library management and advanced query capabilities
 * 
 * @author AuraBoot Team
 * @since 2.2.0
 */
public interface FieldLibraryService {

    /**
     * List all fields grouped by semantic type
     * Groups fields by their semantic type for library view
     * 
     * @return Map of semantic type to field list
     */
    Map<String, List<MetaFieldDTO>> listFieldsBySemanticType();

    /**
     * Search fields with advanced filters
     * Supports multi-dimensional filtering including usage count
     * 
     * @param request Search request with filters
     * @return Paginated field results
     */
    PageResult<MetaFieldDTO> searchFields(FieldSearchRequest request);

    /**
     * Get field recommendations for model binding
     * Recommends fields based on semantic type similarity and usage frequency
     * 
     * @param modelPid Model PID
     * @param semanticType Semantic type filter (optional)
     * @return List of recommended fields with relevance scores
     */
    List<FieldRecommendation> getFieldRecommendations(String modelPid, String semanticType);

    /**
     * Get system fields
     * Returns predefined system fields (id, created_at, updated_at, tenant_id, deleted)
     * 
     * @return List of system fields
     */
    List<MetaFieldDTO> getSystemFields();

    /**
     * Get commonly used business fields
     * Returns fields with high usage count
     * 
     * @return List of common business fields
     */
    List<MetaFieldDTO> getCommonBusinessFields();

    /**
     * Get unused fields
     * Returns fields with zero usage count
     * 
     * @return List of unused fields
     */
    List<MetaFieldDTO> getUnusedFields();
}

File: web-admin/app/plugins/core-designer/components/studio/services/fields/FieldLibraryService.ts
import { get, post } from '~/shared/services/http-client';
import type {
  MetaFieldDTO,
  FieldSearchRequest,
  FieldRecommendation,
} from '~/plugins/core-designer/components/studio/workbench/panels/fields/types';

/**
 * Field Library Service.
 * Provides API calls to backend field library and model field endpoints.
 *
 * @since 3.1.0
 */
export class FieldLibraryService {
  /**
   * List all fields grouped by semantic type.
   * GET /api/meta/field-library
   */
  async listBySemanticType(): Promise<Record<string, MetaFieldDTO[]>> {
    const result = await get<Record<string, MetaFieldDTO[]>>('/api/meta/field-library');
    return result?.data ?? {};
  }

  /**
   * Search fields with advanced filters.
   * POST /api/meta/field-library/search
   */
  async search(request: FieldSearchRequest): Promise<{ records: MetaFieldDTO[]; total: number }> {
    const result = await post<{ records: MetaFieldDTO[]; total: number }>(
      '/api/meta/field-library/search',
      request,
    );
    return result?.data ?? { records: [], total: 0 };
  }

  /**
   * Get field recommendations for a model.
   * GET /api/meta/field-library/recommendations?modelPid=xxx
   */
  async getRecommendations(
    modelPid: string,
    semanticType?: string,
  ): Promise<FieldRecommendation[]> {
    const params: Record<string, string> = { modelPid };
    if (semanticType) params.semanticType = semanticType;
    const result = await get<FieldRecommendation[]>(
      '/api/meta/field-library/recommendations',
      params,
    );
    return result?.data ?? [];
  }

  /**
   * Get model-bound fields.
   * GET /api/meta/models/{modelPid}/fields
   */
  async getModelFields(modelPid: string): Promise<MetaFieldDTO[]> {
    const result = await get<MetaFieldDTO[]>(`/api/meta/models/${modelPid}/fields`);
    return result?.data ?? [];
  }

  /**
   * Get system fields.
   * GET /api/meta/field-library/system-fields
   */
  async getSystemFields(): Promise<MetaFieldDTO[]> {
    const result = await get<MetaFieldDTO[]>('/api/meta/field-library/system-fields');
    return result?.data ?? [];
  }

  /**
   * Get common business fields.
   * GET /api/meta/field-library/common-fields
   */
  async getCommonFields(): Promise<MetaFieldDTO[]> {
    const result = await get<MetaFieldDTO[]>('/api/meta/field-library/common-fields');
    return result?.data ?? [];
  }
}

export const fieldLibraryService = new FieldLibraryService();
